### DISCUSSION TITLE
How to attach debugger

### DISCUSSION THREAD

**Discussion by ptvbnk** [cake-build/cake#3382] [Q&A] [answered]
Hi,
I am trying to contribute to this project - #2443 
I am able to setup and successfully run the stuff on getting started [page](https://cakebuild.net/docs/getting-started/setting-up-a-new-project).

I am having problem with debugging cake library from the other repo which uses .cake file to copy files like mentioned in the issue.
Basically, I am using .cake file to move files from one folder to another without any link to dotnet code in the repo.
I searched on internet and couldn't find attaching debugger for this use case.
Please guide with some resources here.

Also, I am unable to recreate the scenario mentioned in the issue. Even the correct case mentioned is not working i.e. copy of files is not happening at all.

Thanks.

**Comment by augustoproiete** [comment, marked as answer]
Hey @coder2213 thanks for the interest in improving https://github.com/cake-build/cake/issues/2443

There's a few different ways to do that... A simple one would be to:

- Run the Cake script with the flag `-debug` e.g. `dotnet cake -debug test.cake`
You should get a message similar to:

```
Attach debugger to process 12345 to continue...
```

- Then, open the Cake solution in Visual Studio
- Put the breakpoints in the places you want (in your case, probably somewhere in `cake/src/Cake.Common/IO/FileCopier.cs`)
- Go to the menu `Debug` :arrow_right: `Attach to Process`, and find the process with the number indicated by running Cake as per above

![image](https://user-images.githubusercontent.com/177608/120117191-5875c100-c162-11eb-8b99-7206c3d04af7.png)

---

Another alternative is to

- Set the `Cake` project as the startup project
![image](https://user-images.githubusercontent.com/177608/120117221-78a58000-c162-11eb-97d1-bb16d6f64994.png)

- Update the properties of the `Cake` project to run the Cake file that you want to debug
![image](https://user-images.githubusercontent.com/177608/120117240-82c77e80-c162-11eb-86c2-46c54cac6673.png)

- Put the breakpoints in the places you want (in your case, probably somewhere in `cake/src/Cake.Common/IO/FileCopier.cs`)

- And run (F5) normally as if you were debugging a Console application

**Reply by ptvbnk** [reply]
Tried first alternative.
Attached debugger the ID provided in CLI when I entered the command.
cake\src\Cake.Core\Scripting\Scriptrunner.cs 
Only able to hit breakpoints in this class at Run method line 62.
But I observed not all lines are getting hit even though they are on next line.

I tried second alternative and got the below error.
I understood that cake\src\Cake\Commands\DefaultCommand.cs file line 40 has a DefaultCommandSettings variable which is not registered.
Should I register it for the sake of debugging code?
```
Autofac.Core.Registration.ComponentNotRegisteredException
  HResult=0x80131500
  Message=The requested service 'Cake.Commands.DefaultCommandSettings' has not been registered. To avoid this exception, either register a component to provide the service, check for service registration using IsRegistered(), or use the ResolveOptional() method to resolve an optional dependency.
  Source=Autofac
  StackTrace:
   at Autofac.ResolutionExtensions.ResolveService(IComponentContext context, Service service, IEnumerable`1 parameters)
   at Autofac.ResolutionExtensions.Resolve(IComponentContext context, Type serviceType)
   at Cake.Infrastructure.Composition.AutofacTypeResolver.Resolve(Type type) in C:\Users\Teja\Desktop\cake\src\Cake\Infrastructure\Composition\AutofacTypeResolver.cs:line 22
   at Spectre.Console.Cli.TypeResolverAdapter.Resolve(Type type) in /_/src/Spectre.Console/Cli/Internal/TypeResolverAdapter.cs:line 26
```
Please guide me on debugging this.

**Reply by augustoproiete** [reply]
@coder2213 Is it possible that the source code you have on you machine don't match the version of Cake you're running?

e.g. If you're running Cake v1.1.0, then the source code you should be looking at would be the source under the [`v1.1.0` tag](https://github.com/cake-build/cake/tree/v1.1.0)

**Reply by ptvbnk** [reply]
I tried moving to v1.1.0 tag code. Still I am not able to hit breakpoints.

**Reply by augustoproiete** [reply]
@coder2213 Have you tried the second method by just passing arguments to the runner when hitting F5?

**Reply by ptvbnk** [reply]
Yes, I did.

I am using below code in other repo where I am trying to use cake (exampleapp solution)

```
var target = Argument("target", "Test");
var configuration = Argument("configuration", "Release");

Task("Clean")
	  .Does(() =>
{
CopyFiles(
    new[]
    {
        new FilePath("./source/file1.txt")
    },
    "target", // Even though this should work, try `MakeAbsolute(Directory("Destination"))`
                   // for full effect
    preserveFolderStructure: true);
}
);
```

Any issue with above script?`

**Reply by augustoproiete** [reply]
Looks like your script doesn't do anything... It just declares a task, but doesn't actually run it.

You're missing a `RunTarget(target);` at the very end.

```cakescript
// Select which target to use based on user input or default to "Test" if not provided by the user
var target = Argument("target", "Test");

// Define the code to execute when the "Clean" target is executed
Task("Clean")
    .Does(() =>
{
    // (...)
});

// (...)

// Runs the selected target defined at the beginning
RunTarget(target);
```

**Reply by ptvbnk** [reply]
Added `RunTarget `and its working.
Thanks.
